plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1236,39 +1236,59 @@ public void executeRemoveOrphanFiles(ConnectorSession session, IcebergTableExecu
-                .collect(toImmutableList());
-        Set<String> validMetadataFiles = concat(manifests.stream(), manifestLists.stream(), otherMetadataFiles.stream())
-                .map(IcebergMetadata::fileName)
-                .collect(toImmutableSet());
-        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, validMetadataFiles, "metadata");
+        for (Snapshot snapshot : table.snapshots()) {
+            if (snapshot.manifestListLocation() != null) {
+                validMetadataFilePaths.add(snapshot.manifestListLocation());
+            }
+
+            for (ManifestFile manifest : snapshot.allManifests(table.io())) {
+                if (validMetadataFilePaths.contains(manifest.path())) {
+                    // Already read this manifest
+                    continue;
+                }
+
+                validMetadataFilePaths.add(manifest.path());
+                if (manifest.content() == ManifestContent.DATA) {
+                    try (ManifestReader<DataFile> manifestReader = ManifestFiles.read(manifest, table.io());
+                        CloseableIterator<DataFile> dataFiles = manifestReader.iterator()) {
+                        while (dataFiles.hasNext()) {
+                            validDataFilePaths.add(dataFiles.next().path().toString());
+                        }
+                    }
+                    catch (IOException e) {
+                        throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Unable to list manifest file content from " + manifest.path(), e);
+                    }
+                }
+                else if (manifest.content() == ManifestContent.DELETES) {
+                    try (ManifestReader<DeleteFile> manifestReader = ManifestFiles.readDeleteManifest(manifest, table.io(), table.specs());
+                            CloseableIterator<DeleteFile> dataFiles = manifestReader.iterator()) {
+                        while (dataFiles.hasNext()) {
+                            validDeleteFilePaths.add(dataFiles.next().path().toString());
+                        }
+                    }
+                    catch (IOException e) {
+                        throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Unable to list manifest file content from " + manifest.path(), e);
+                    }
+                }
+                else {
+                    throw new TrinoException(ICEBERG_INVALID_METADATA, "Unknown Iceberg manifest content: " + manifest);
+                }
+            }
+        }
+
+        validMetadataFilePaths.addAll(metadataFileLocations(table, false));
+        validMetadataFilePaths.add(versionHintLocation(table));
+
+        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, union(validDataFilePaths.build(), validDeleteFilePaths.build()), "data");
+        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, validMetadataFilePaths, "metadata");
```

Comment:
is it an error to have a data file in a metadata directory (or vice versa)?
is table corruption a fair punishment?

i guess this is no behavioral change here in this PR, so no change requested. But also, since you collect full paths, it seems easy to change if we want to

Reply:
I'm going to leave it for now, but if we support more of the Iceberg write location properties / LocationProvider implementations we'll need to change this.